File: speeder.py
import time

def log(s):
    print(s)
    L.write(s)
    L.write('\n')

CNT = 1000      # number of runs

L = open('speeder.log', 'a')

m0, m1, s = 100_000_000., -1, 0    # min, max, total

log(f'{"min":^19} | {"max":^19} | {"avg":^19} | step')

for r in range(CNT):
    t = time.time()
    a, b, c, d = 1, 0, 1.1, 1.01
    for i in range(50000):
        a = (a * 123_456_789) % 987_654_321
        if a % 2: b = b + 1
        c = c / d
        if c < 0.1: d = 0.99
        if c > 9.9: d = 1.01
    t = time.time() - t
    s = s + t
    if t < m0: m0 = t
    if t > m1: m1 = t

    log(f'{m0:19.15f} | {m1:19.15f} | {s/(r+1):19.15f} | {r+1}')

open('finetest.py', 'a').write(f'SCALE = {(s-m0-m1) / (CNT-2) * 100}\n')

File: speeder.cpp
#include <iostream>
#include <fstream>
#include <iomanip>
#include <chrono>

#define CNT 1000        // number of runs

void log(const std::string& s, std::ofstream& logFile) {
    std::cout << s << std::endl;
    logFile << s << std::endl;
}

int main() {
    std::ofstream logFile("speeder.log", std::ios::app);
    if (!logFile.is_open()) {
        std::cerr << "Failed to open speeder.log" << std::endl;
        return 1;
    }

    double m0 = 100000000.0, m1 = -1.0, s = 0.0;    // min, max, total

    logFile << std::setw(19) << "min" << " | "
            << std::setw(19) << "max" << " | "
            << std::setw(19) << "avg" << " | step" << std::endl;

    for (int r = 0; r < CNT; ++r) {
        auto start = std::chrono::high_resolution_clock::now();

        long long a = 1, b = 0;
        double c = 1.1, d = 1.01;

        for (int i = 0; i < 50000; ++i) {
            a = (a * 123456789) % 987654321;
            if (a % 2) b = b + 1;
            c = c / d;
            if (c < 0.1) d = 0.99;
            if (c > 9.9) d = 1.01;
        }

        std::chrono::duration<double> elapsed = std::chrono::high_resolution_clock::now() - start;

        double t = elapsed.count();     // Get elapsed time in seconds
        s += t;
        if (t < m0) m0 = t;
        if (t > m1) m1 = t;

        logFile << std::fixed << std::setprecision(15)
                << std::setw(19) << m0 << " | "
                << std::setw(19) << m1 << " | "
                << std::setw(19) << s / (r + 1) << " | " << r + 1 << std::endl;
    }
    
    logFile << std::endl;
    logFile.close();

    std::ofstream fineTestFile("finetest.py", std::ios::app);
    if (!fineTestFile.is_open()) {
        std::cerr << "Failed to open finetest.py" << std::endl;
        return 1;
    }

    fineTestFile << "SCALE_C = " 
                 << (s - m0 - m1) / (CNT - 2) * 1000 << std::endl;

    fineTestFile.close();
}